### ISSUE TITLE
Document 'heroku:16' vs 'heroku:16-build' docker images

### ISSUE THREAD

**Issue by edmorley** [heroku/stack-images#54]
The recent [Heroku-16 blog post](https://blog.heroku.com/heroku-16-is-generally-available) and the [Heroku-16 stack](https://devcenter.heroku.com/articles/heroku-16-stack#heroku-16-docker-image) / [Container Registry and Runtime](https://devcenter.heroku.com/articles/container-registry-and-runtime#stack-images) devcenter pages just mention the `heroku/heroku:16` docker image.

However the [docker hub tags page](https://hub.docker.com/r/heroku/heroku/tags/) lists two variants, `16` and `16-build`.

From reading #53 (not yet merged), `16-build` is the image that contains all of the development headers, so is more likely the image that people will need when developing locally, rather than the base `16` image.

Could the docs be updated to clarify this? (The two deventer pages, this repo's readme, and the docker hub description.)

Many thanks :-)

**Comment by edmorley** [comment]
@tt any thoughts on this? :-)

**Comment by tt** [comment]
Our documentation should definitely be updated.

The build variant is what we use for our build system. We need more packages to build things than to run them.

It's true that for building things locally, you need the build variant too but [Docker's multi-stage builds](https://docs.docker.com/engine/userguide/eng-image/multistage-build/) will let you still produce small runtime images.

```dockerfile
FROM heroku/heroku:16-build as build
...

FROM heroku/heroku:16
COPY --from=build <path> .
...
```

**Comment by tt** [comment]
I tried building two internal apps in various configurations:

**App A:**

| | Before multi-stage | After multi-stage | After multi-stage with build stack |
| --- | ---: | ---: | ---: |
| Cedar-14 | 1.84 GB | 1.61 GB | *N/A* |
| Heroku-16 | 1.31 GB | 1.09 GB | 698 MB |

**App B:**

| | Before multi-stage | After multi-stage | After multi-stage with build stack |
| --- | ---: | ---: | ---: |
| Cedar-14 | 1.72 GB | 1.5 GB | *N/A* |
| Heroku-16 | 1.19 GB | 982 MB | 593 MB |

**Comment by jbyrum** [comment]
I've updated the heroku-16 documentation to mention the build image, without promoting it across dev center. I don't agree that the build variant is what most customer will need, as they are typically installing one or two packages from apt:

https://devcenter.heroku.com/articles/heroku-16-stack#heroku-16-docker-image

**Comment by edmorley** [comment]
> It's true that for building things locally, you need the build variant too but Docker's multi-stage builds will let you still produce small runtime images.

Yeah great idea - thank you!

> I don't agree that the build variant is what most customer will need, as they are typically installing one or two packages from apt

I wasn't suggesting that the build image necessarily be promoted over the other - just that it at least be explained somewhere so people could decide themselves, which has now been done - thank you :-)

I think also that we're approaching this from different use-cases. My issue was filed about using these images for local development parity with heroku (rather than say for use with the container registry and runtime) - which also means we also have to emulate the buildpack compile that runs under the build image. 

For example our Python project needs libmysqlclient-dev, libmemcached-dev and zlib1g-dev, just to successfully run the `pip install` performed during buildpack compile. Now we could of course use the smaller runtime heroku-16 image, and manually install the development header packages ourselves, however there's no guarantee that our image will match `heroku:16-build` (for example it could contain different versions depending on when it was updated), and so a safer option would be to use the build image itself - ideally via the multi-stage build method mentioned above. 

That said, the build image is a whopping 334 MB / 855 MB uncompressed (compare this to the base ubuntu:16.04 image's 47 MB / 119 MB uncompressed), so takes an age to pull as part of a Travis CI run (where caching options are limited) - so even we're considering ditching it in favour of the runtime image. On that topic, I've spotted a few size wins for both the runtime and build heroku-16 images that I'll open an issue about soon.
